### PR TITLE
Gradle Module Metadata: fail publication when no version information

### PR DESCRIPTION
**Pull request by ljacomet**
With this change, it becomes illegal to create a Gradle Module Metadata
file that has depedencies or constraints declared without any version at
 all across all variants.

### CODE REVIEW

File: subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishDependenciesIntegTest.groovy [hunk 1]
```diff
@@ -80,6 +80,7 @@ class MavenPublishDependenciesIntegTest extends AbstractMavenPublishIntegTest {
 
             dependencies {
                 api $dependencyNotation
+                api 'group:projectB:1.0'
```
Comment (melix) [inline subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishDependenciesIntegTest.groovy]: :o: why this change?

Comment (ljacomet) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishDependenciesIntegTest.groovy]: Because otherwise this test triggers the new check: it has no version anywhere

Comment (melix) [reply subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishDependenciesIntegTest.groovy]: Ah I see, this tests with _no version_

**Comment by ljacomet** [comment]
@melix test added
